### PR TITLE
Pull image if the image doesn't exist

### LINKED ISSUE #272: Shiplift client doesn't automatically pull image

### PR DESCRIPTION
**Pull request by tony84727**
Fix #272 
Pull the image when creating container returns 404 not found.

### CODE REVIEW

File: src/clients/http.rs [hunk 1]
```diff
@@ -110,13 +113,32 @@ impl Http {
...
-            .create(&options_builder.build())
-            .await;
-        let id = create_result.unwrap().id;
+        let create_result = self.create_container(&options_builder).await;
+        let id = {
+            match create_result {
+                Ok(container) => container.id,
+                Err(err) => match err {
+                    shiplift::Error::Fault { code, message: _ } => {
+                        if code != StatusCode::NOT_FOUND {
```
Comment (thomaseizinger) [inline src/clients/http.rs:122]: Can we merge this pattern matching into one and thereby reduce indentation for this logic?

I am thinking of something like:

```rust
match create_result {
	Ok(c) => c.id,
	Err(shiplift::Error::Fault { code, .. }) if code == 404 => {
		// do image pull logic
	}
	e => panic!("Failed to create container: {}", e)
}
```

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Thank you for contributing!

I left two small comments inline.

Additionally, it would be really great if we could somehow cover this in our integration tests. I am wondering how https://github.com/testcontainers/testcontainers-rs/blob/dev/tests/http_client.rs even works at the moment. Is `hello-world` simply present on the GitHub actions dev machines?

File: src/clients/http.rs (hunk 1 above)
Comment (tony84727) [reply src/clients/http.rs]: Thank you! I don't even know we can write it like that. Fixed.

[the remark this answers is not in the document]
Comment (tony84727) [reply src/clients/http.rs]: Fixed

**Comment by tony84727** [comment]
> I am wondering how https://github.com/testcontainers/testcontainers-rs/blob/dev/tests/http_client.rs even works at the moment. Is hello-world simply present on the GitHub actions dev machines?

I guess it's because of another integration test:
https://github.com/testcontainers/testcontainers-rs/blob/e36491f7e6fc84e0de15be148f87f1ded06e0f41/tests/podman.rs#L4-L7

docker cli will pull the image if the engine returns an image not found error.

> Additionally, it would be really great if we could somehow cover this in our integration tests.

Yes, but I'm not sure how to write the test for this case, as the test will need to be run with a clean docker engine.

Maybe we write an integration test that will delete the image at the beginning?

Another option might be using [docker in docker, dind](https://hub.docker.com/_/docker). Not sure if we can run dind by Github action though.

**Comment by thomaseizinger** [comment]
> I guess it's because of another integration test:

Right, that makes sense!

> Yes, but I'm not sure how to write the test for this case, as the test will need to be run with a clean docker engine.
> Maybe we write an integration test that will delete the image at the beginning?

If we do that, we need make sure things run in sequence properly, otherwise we might be subject to race conditions. cargo will run all integration test files in sequence, so we can use that to have a dedicated test that uses shiplift's API to first delete the `hello-world` docker image and then use testcontainers to start it up.

**Comment by tony84727** [comment]
> cargo will run all integration test files in sequence, so we can use that to have a dedicated test that uses shiplift's API to first delete the `hello-world` docker image and then use testcontainers to start it up.

Okay. added an integration that will delete `hello-world` image at the beginning and then test if we can spin up a `hello-world` container.

**Comment by thomaseizinger** [comment]
> Okay. added an integration that will delete `hello-world` image at the beginning and then test if we can spin up a `hello-world` container.

Amazing, thank you!

A final request: can we squash all 4 commits into a single one please?

**Comment by tony84727** [comment]
> A final request: can we squash all 4 commits into a single one please?

Sure, feel free to squash those commits. Should I squash commits then force push? Or you will use Github's squash merge?

**Comment by thomaseizinger** [comment]
> Sure, feel free to squash those commits. Should I squash commits then force push? Or you will use Github's squash merge?

Wasn't planning on using squash merge but can do if you are okay with it. You will lose the signatures on your commits but I guess you are fine with that given that you are suggesting the option? :)

Just noticed that CI is not passing. Sorry for the late notice, this new feature of having to activate CI checks every time is annoying ...

**Comment by tony84727** [comment]
> Wasn't planning on using squash merge but can do if you are okay with it. You will lose the signatures on your commits but I guess you are fine with that given that you are suggesting the option? :)

I think it's okay

> Just noticed that CI is not passing. Sorry for the late notice, this new feature of having to activate CI checks every time is annoying ...

CI seems to be failing on the `dev` branch. I've opened
https://github.com/testcontainers/testcontainers-rs/pull/289 to fix it.

Will deal with the formatting after the CI is fixed.

**Comment by tony84727** [comment]
> CI seems to be failing on the dev branch. I've opened
#289 to fix it.
>
> Will deal with the formatting after the CI is fixed.

Fixed CI and formatting issue.

**Comment by thomaseizinger** [comment]
Thank you!

**Pull request merged**